### DISCUSSION TITLE
React on event from other tasmota device

### DISCUSSION THREAD

**Discussion by ChrisVonFrosta** [arendst/Tasmota#20021] [Q&A]
i have a berry script that actually can handle switch inputs, pwm outputs and relays.
i would like to use other tasmota devices as "wifi sensors" like PIR sensors. What is the best way to get the events on my main Tasmota device (in Berry script)? i have no mqtt broker in network and i want not to have one, unless it could run on main Tasmota device.

thanks a lot!

**Comment by barbudor** [comment]
I don't think a MQTT broker would ever be added to Tasmota, although there are a couple of such solution around in github

If MQTT is out you back to basics such as HTTP or UDP
I would consider HTTP a bit heavy but you can join that [discussion] (https://github.com/arendst/Tasmota/discussions/20010#discussioncomment-7568588) where 'tsmoneta' user came with a smart solution using the syslog feature.
Your main ESP32 could receive the logs from other tasmota using UDP receiver and parse them
Still you will have to experiment to see if Berry can cope with receiving UDP messages from many devices
Alternatively if all of your sensors run Berry, instead of using syslog you could just implement your own UDP based protocol

**Reply by sfromis** [reply]
Well, syslog is not really feasible for device-to-device communication, until someone decides to do a project with implementing a syslog server for Tasmota. Without MQTT, the simplest (least amount of coding) is probably using http. Sensor devices can use WebQuery to invoke commands on the "main Tasmota device", but doing that with multiple fields gets to be a mess if using rules. Polling is simpler, but a high poll frequency is not always so great.

Note that the http support in Tasmota becomes unreliable if same device can get more than one request at a time.

**Reply by barbudor** [reply]
> until someone decides to do a project with implementing a syslog server

Berry udp ?

**Reply by ChrisVonFrosta** [reply]
at first, thank you!
http was my first think but i have no initial starting point
i took a look to [tcpserver class ](https://tasmota.github.io/docs/Berry/#tcpserver-class) and [udp class](https://tasmota.github.io/docs/Berry/#udp-class)
i do not really know the difference between both. But on each of them i have to poll to get the message?
i could imaging a solution with a webserver on the main/master ESP32.

what is about this: in tasmote i can do (never did now) declare functions as tasmota commands and i can execute command remotly over http.
So, if I have a function called "sensor_input(id,value)" on the master ESP32, which is declared as a command, the sensor client should be able to trigger the master over HTTP, right?

**Reply by sfromis** [reply]
Sure, it should not be that hard to create a syslog server, and untangle the message format. Just working out the nitty-gritty.

**Reply by sfromis** [reply]
Using `tasmota.add_cmd` you can register a Tasmota command name, with a function called when Tasmota receives that command.
See https://tasmota.github.io/docs/Berry/#functions-to-create-custom-tasmota-command
Function arguments are (cmd, idx, payload, raw)

Such a command can be called via console, MQTT, http, serial, rules, and so on.

**Reply by joba-1** [reply]
> i took a look to [tcpserver class ](https://tasmota.github.io/docs/Berry/#tcpserver-class)and [udp class](https://tasmota.github.io/docs/Berry/#udp-class)
> i do not really know the difference between both.

The difference between tcp and udp is that tcp maintains a connection and received data is acknowledged. I.e. you will know if an error has happened. Udp is fire and forget.

And yes, you need to poll.
